### PR TITLE
Fix cmd integration tests to cleanup plugin directory

### LINKED ISSUE #1185: Fix cmd integration tests to cleanup plugin directory
# Expected Behavior

Unexpected exits in any of the OPA tests should not result in generated files being left around in the working copy (or elsewhere).

# Actual Behavior

Currently, if the cmd tests are exited unexpectedly, e.g., due to the developer ctrl+c-ing out in the middle, the plugin tests leave a bunch of files on disk.

# Steps to Reproduce the Problem

  1. Run `go test -v ./cmd`  
  2. Ctrl+c after 1 second
  3. Run `git status` to see dirty files

For example:

```
(add-eval-coverage)torin:~/go/src/github.com/open-policy-agent/opa$ go test ./cmd  -v
^Csignal: interrupt
FAIL	github.com/open-policy-agent/opa/cmd	1.124s
(add-eval-coverage)[*]torin:~/go/src/github.com/open-policy-agent/opa$ git status
On branch add-eval-coverage
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	cmd/plugin_test_tempdir435644132/

```

# Additional Info

### PR DESCRIPTION
**Pull request by repenno**
Fixes #1185

Signed-off-by: repenno <rapenno@gmail.com>

### CODE REVIEW

File: cmd/plugins_test.go [hunk 1]
```diff
@@ -28,24 +31,59 @@ import (
 var initChan = make(chan struct{}, 256)
 var testDirRoot string
 
+const (
+	rootDir   string = "./"
```
Comment (srenatus) [inline cmd/plugins_test.go]: [nit] Do we _need_ the `string`?
```suggestion
	rootDir   = "./"
```

File: cmd/plugins_test.go [hunk 2]
```diff
@@ -28,24 +31,59 @@ import (
...
+// Code is duplicated from MakeTempFS() due to https://github.com/open-policy-agent/opa/issues/1185
+func makeDirWithSharedObjects(files map[string]string, ext string) (tempRootDir string, cleanup func()) {
+
+	tempRootDir, err := ioutil.TempDir(rootDir, prefixDir)
 	if err != nil {
 		panic(err)
 	}
+	cleanup = func() {
+		if err := os.RemoveAll(tempRootDir); err != nil {
+			fmt.Println("Failed to cleanup directory: ", tempRootDir)
```
Comment (srenatus) [inline cmd/plugins_test.go]: ❓ Would it be interesting to include `err`? If so, we could do
```suggestion
			fmt.Println("Failed to cleanup directory %q: %v", tempRootDir, err)
```
if not, we could change the `if` to
```go
if os.RemoveAll(tempRootDir) != nil {
```

File: cmd/plugins_test.go [hunk 3]
```diff
@@ -28,24 +31,59 @@ import (
...
+)
+
 // makeDirWithSharedObjects creates a new temporary directory containing files under the runtime directory
 // it compiles all .go files into shared object files with extension ext in the corresponding directory
 // It returns the root of the directory and a cleanup function.
-func makeDirWithSharedObjects(files map[string]string, ext string) (root string, cleanup func()) {
-	root, cleanup, err := test.MakeTempFS("./", "plugin_test_tempdir", files)
+//
+// Code is duplicated from MakeTempFS() due to https://github.com/open-policy-agent/opa/issues/1185
+func makeDirWithSharedObjects(files map[string]string, ext string) (tempRootDir string, cleanup func()) {
```
Comment (srenatus) [inline cmd/plugins_test.go]: [nit] Do we need the named return arguments? This function is quite long now, and the naked `return` makes you go hunt for what it potentially has returned... How about
```suggestion
func makeDirWithSharedObjects(files map[string]string, ext string) (string, func()) {
```
and explicitly returning values?

File: cmd/plugins_test.go [hunk 4]
```diff
@@ -28,24 +31,59 @@ import (
...
+	// We install the signal handler soon after the creation of the temp directory
+	signalHandler(tempRootDir, cleanup)
+
+	for path, content := range files {
+		dirname, filename := filepath.Split(path)
+		dirPath := filepath.Join(tempRootDir, dirname)
+		fmt.Println("dirpath", dirPath)
+		if err := os.MkdirAll(dirPath, 0777); err != nil {
+			fmt.Println("err", err)
+			panic(err)
```
Comment (srenatus) [inline cmd/plugins_test.go]: 💭 there's quite a few `panic()`s in here -- would it make sense to add an `error` return instead? What do you think? 😃 
Alternatively, we could pass in `*testing.T` and call `t.Fatal`. 🤔

File: cmd/plugins_test.go [hunk 5]
```diff
@@ -28,24 +31,59 @@ import (
...
+			fmt.Println("Failed to cleanup directory: ", tempRootDir)
+		}
+	}
+	// We install the signal handler soon after the creation of the temp directory
+	signalHandler(tempRootDir, cleanup)
+
+	for path, content := range files {
+		dirname, filename := filepath.Split(path)
+		dirPath := filepath.Join(tempRootDir, dirname)
+		fmt.Println("dirpath", dirPath)
```
Comment (srenatus) [inline cmd/plugins_test.go]: 🧹 I suppose this can be cleaned up?

File: cmd/plugins_test.go [hunk 6]
```diff
@@ -313,3 +348,18 @@ func TestLambdaFileWalker(t *testing.T) {
...
+
+func signalHandler(tempRootDir string, cleanup func()) {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		s := <-signalChan
+		switch s {
+
+		case syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT:
+			fmt.Println("Received an ", s.String(), "cleaning ", tempRootDir, "and exiting")
```
Comment (srenatus) [inline cmd/plugins_test.go]: [nit] What would you say to using `fmt.Printf` instead? 😉

File: cmd/plugins_test.go (hunk 2 above)
Comment (repenno) [reply cmd/plugins_test.go]: Done but used fmt.Printf

File: cmd/plugins_test.go (hunk 3 above)
Comment (repenno) [reply cmd/plugins_test.go]: There are really no guidelines in OPA as long a function should be, but I found this one:

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

So, fixed it.

File: cmd/plugins_test.go (hunk 1 above)
Comment (repenno) [reply cmd/plugins_test.go]: Done

File: cmd/plugins_test.go (hunk 4 above)
Comment (repenno) [reply cmd/plugins_test.go]: This is existing code. When fixing a bug try to leave in place anything out of scope. We can open another issue to fix these.

File: cmd/plugins_test.go (hunk 6 above)
Comment (repenno) [reply cmd/plugins_test.go]: This could be in a OPA style guide.

File: cmd/plugins_test.go (hunk 4 above)
Comment (srenatus) [reply cmd/plugins_test.go]: Yup fair enough. Sorry I hadn't checked that 😃

**Comment by tsandall** [comment]
@repenno looks good. Do you want to squash your commits yourself or would you like me to do it for you?

**Pull request merged**